File: templates/Enterprise-Template/src/csharp/EnterpriseBotTemplate/Bot Framework/Extensions/ActivityExtensions.cs
using Microsoft.Bot.Connector;
using Microsoft.Bot.Schema;
using System;
using System.Linq;

namespace $safeprojectname$.Extensions
{
    public static class ActivityExtensions
    {
        public static bool IsStartActivity(this Activity activity)
        {
            switch (activity.ChannelId)
            {
                case Channels.Cortana:
                    {
                        var intent = activity.Entities?.FirstOrDefault(entity => entity.Type.Equals("Intent", StringComparison.Ordinal));
                        var intentName = string.Empty;
                        if (intent != null)
                        {
                            intentName = intent.Properties["name"]?.ToString();
                        }

                        return (intent != null
                                && !string.IsNullOrEmpty(intentName)
                                && intentName == "Microsoft.Launch");
                    }

                case Channels.Skype:
                    {
                        if (activity.Type == ActivityTypes.ContactRelationUpdate && activity.Action == "add")
                        {
                            return true;
                        }

                        return false;
                    }

                case Channels.Directline:
                case Channels.Emulator:
                case Channels.Webchat:
                case Channels.Msteams:
                    {
                        if (activity.Type == ActivityTypes.ConversationUpdate)
                        {
                            // When bot is added to the conversation (triggers start only once per conversation)
                            if (activity.MembersAdded.Any(m => m.Id == activity.Recipient.Id))
                            {
                                return true;
                            }
                        }

                        return false;
                    }

                default:
                    {
                        return false;
                    }
            }
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/extensions/activityExtensions.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    Activity,
    ActivityTypes,
    ChannelAccount,
    ConversationReference } from 'botbuilder';

export namespace ActivityExtensions {
    export function createReply(source: Activity, text?: string, local?: string): Activity {
        const reply: string = text || '';

        return {
            channelId: source.channelId,
            conversation: source.conversation,
            from: source.recipient,
            label: source.label,
            locale: local,
            recipient: source.from,
            replyToId: source.id,
            serviceUrl: source.serviceUrl,
            text: reply,
            timestamp: new Date(),
            type: ActivityTypes.Message,
            valueType: source.valueType,
            localTimezone: source.localTimezone,
            listenFor: source.listenFor,
            semanticAction: source.semanticAction
        };
    }

    export function getContinuationActivity(source: Partial<ConversationReference>): Partial<Activity> {
        if (!source) {
            throw new Error('source needs to be defined');
        }

        return {
            type: ActivityTypes.Event,
            name: 'ContinueConversation',
            channelId: source.channelId,
            serviceUrl: source.serviceUrl,
            conversation: source.conversation,
            recipient: source.bot,
            from: source.user,
            relatesTo: <ConversationReference> source
        };
    }

    export function isStartActivity(activity: Activity): boolean {
        switch (activity.channelId) {
            case 'skype': {
                if (activity.type === ActivityTypes.ContactRelationUpdate && activity.action === 'add') {
                    return true;
                }

                return false;
            }
            case 'directline':
            case 'emulator':
            case 'webchat':
            case 'msteams': {
                if (activity.type === ActivityTypes.ConversationUpdate) {
                    if (activity.membersAdded && activity.membersAdded.some((m: ChannelAccount) => m.id === activity.recipient.id)) {
                        // When bot is added to the conversation (triggers start only once per conversation)
                        return true;
                    }
                }

                return false;
            }
            default:
                return false;
        }
    }
}
